File: opensrp-chw-referral/src/main/java/org/smartregister/chw/referral/interactor/BaseIssueReferralInteractor.java
package org.smartregister.chw.referral.interactor;

import androidx.annotation.VisibleForTesting;

import com.google.gson.Gson;
import com.nerdstone.neatformcore.domain.model.NFormViewData;

import org.apache.commons.lang3.text.WordUtils;
import org.jetbrains.annotations.NotNull;
import org.json.JSONObject;
import org.smartregister.chw.referral.ReferralLibrary;
import org.smartregister.chw.referral.contract.BaseIssueReferralContract;
import org.smartregister.chw.referral.domain.ReferralTask;
import org.smartregister.chw.referral.util.AppExecutors;
import org.smartregister.chw.referral.util.Constants;
import org.smartregister.chw.referral.util.JsonFormConstants;
import org.smartregister.chw.referral.util.JsonFormUtils;
import org.smartregister.chw.referral.util.ReferralUtil;
import org.smartregister.chw.referral.util.Util;
import org.smartregister.repository.AllSharedPreferences;

import java.util.HashMap;
import java.util.Objects;
import java.util.UUID;

import timber.log.Timber;

public class BaseIssueReferralInteractor implements BaseIssueReferralContract.Interactor {

    private AppExecutors appExecutors;

    @VisibleForTesting
    private BaseIssueReferralInteractor(AppExecutors appExecutors) {
        this.appExecutors = appExecutors;
    }

    public BaseIssueReferralInteractor() {
        this(new AppExecutors());
    }

    @Override
    public void onDestroy(boolean isChangingConfiguration) {
//        implement
    }

    @Override
    public void saveRegistration(String baseEntityId, HashMap<String, NFormViewData> valuesHashMap, JSONObject jsonObject, final BaseIssueReferralContract.InteractorCallBack callBack) {
        try {
            saveRegistration(baseEntityId, valuesHashMap, jsonObject);
        } catch (Exception e) {
            Timber.e(e);
        }
        callBack.onRegistrationSaved();
    }

    @VisibleForTesting
    void saveRegistration(String baseEntityId, HashMap<String, NFormViewData> valuesHashMap, JSONObject jsonObject) throws Exception {

        AllSharedPreferences allSharedPreferences = ReferralLibrary.getInstance().context().allSharedPreferences();
        ReferralTask referralTask = JsonFormUtils.processJsonForm(allSharedPreferences, baseEntityId, valuesHashMap, jsonObject, Constants.EVENT_TYPE.REGISTRATION);
        referralTask.setGroupId(((NFormViewData) Objects.requireNonNull(valuesHashMap.get("chw_referral_hf")).getValue()).getMetadata().get("openmrs_entity_id").toString());
        referralTask.setFocus(WordUtils.capitalize(jsonObject.getString(JsonFormConstants.REFERRAL_TASK_FOCUS)));

        String problemsStrings = extractReferralProblems(valuesHashMap);
        referralTask.setReferralDescription(problemsStrings);
        Objects.requireNonNull(referralTask.getEvent()).setEventId(UUID.randomUUID().toString());

        Timber.i("Referral Event = %s", new Gson().toJson(referralTask));

        Util.processEvent(allSharedPreferences, referralTask.getEvent());

        ReferralUtil.createReferralTask(referralTask, allSharedPreferences);
    }

    @NotNull
    @SuppressWarnings("unchecked")
    private String extractReferralProblems(HashMap<String, NFormViewData> valuesHashMap) {
        HashMap<String, NFormViewData> valuesMap =
                (HashMap<String, NFormViewData>) valuesHashMap.get(JsonFormConstants.PROBLEM).getValue();

        StringBuilder problemNamesStringBuilder = new StringBuilder();
        for (String key : valuesMap.keySet()) {
            problemNamesStringBuilder.append(valuesMap.get(key).getValue());
            problemNamesStringBuilder.append(", ");
        }

        String problemsStrings = problemNamesStringBuilder.toString();
        problemsStrings = problemsStrings.substring(0, problemsStrings.length() - 1);
        return problemsStrings;
    }

}

File: opensrp-chw-referral/src/main/java/org/smartregister/chw/referral/interactor/BaseIssueReferralInteractor.kt
package org.smartregister.chw.referral.interactor

import com.google.gson.Gson
import com.nerdstone.neatformcore.domain.model.NFormViewData
import org.apache.commons.lang3.text.WordUtils
import org.json.JSONObject
import org.koin.core.inject
import org.smartregister.chw.referral.ReferralLibrary
import org.smartregister.chw.referral.contract.BaseIssueReferralContract
import org.smartregister.chw.referral.domain.ReferralTask
import org.smartregister.chw.referral.util.Constants
import org.smartregister.chw.referral.util.JsonFormConstants
import org.smartregister.chw.referral.util.JsonFormUtils
import org.smartregister.chw.referral.util.LocationUtils
import org.smartregister.chw.referral.util.ReferralUtil.createAddoLinkageTask
import org.smartregister.chw.referral.util.ReferralUtil.createReferralTask
import org.smartregister.chw.referral.util.Util.extractReferralProblems
import org.smartregister.chw.referral.util.Util.processEvent
import timber.log.Timber
import java.util.*

/**
 * This interactor class provides actual implementations for all the functionality used in the
 * Referral forms, it implements [BaseIssueReferralContract.Interactor]
 */
open class BaseIssueReferralInteractor : BaseIssueReferralContract.Interactor {

    private val referralLibrary by inject<ReferralLibrary>()


    @Throws(Exception::class)
    override fun saveRegistration(
            baseEntityId: String, valuesHashMap: HashMap<String, NFormViewData>,
            jsonObject: JSONObject, callBack: BaseIssueReferralContract.InteractorCallBack, isAddoLinkage: Boolean
    ) {
        val allSharedPreferences = referralLibrary.context.allSharedPreferences()
        val extractReferralProblems = extractReferralProblems(valuesHashMap)
        val hasProblems = !extractReferralProblems.isNullOrEmpty()
        if (hasProblems) {
            val referralTask: ReferralTask =
                    JsonFormUtils.processJsonForm(
                            referralLibrary, baseEntityId, valuesHashMap,
                            jsonObject, Constants.EventType.REGISTRATION
                    )

            referralTask.apply {
                groupId = if (isAddoLinkage) LocationUtils.getWardId()
                else (valuesHashMap[JsonFormConstants.CHW_REFERRAL_HF]?.value as NFormViewData?)
                    ?.metadata?.get(JsonFormConstants.OPENMRS_ENTITY_ID)
                    .toString()
                focus = WordUtils.capitalize(jsonObject.getString(JsonFormConstants.REFERRAL_TASK_FOCUS))
                referralDescription = extractReferralProblems
                event.eventId = UUID.randomUUID().toString()
            }

            Timber.i("Referral Event = %s", Gson().toJson(referralTask))

            processEvent(referralLibrary, referralTask.event)

            if (isAddoLinkage) {
                createAddoLinkageTask(referralTask, referralLibrary)
            } else {
                createReferralTask(referralTask, referralLibrary)
            }
        }
        callBack.onRegistrationSaved(hasProblems, isAddoLinkage)
    }

}